### PR TITLE
feat: FIP-8 verifications for contract wallets

### PR DESCRIPTION
**Pull request by horsefacts**
## Motivation

Support verifications signed by smart contracts. See [FIP-8](https://github.com/farcasterxyz/protocol/discussions/109) for more details.

## Change Summary

Update validations to verify smart contract signatures provided in verification messages using [`publicClient.verifyTypedData`](https://viem.sh/docs/actions/public/verifyTypedData.html). This function makes an RPC call to verify smart contract signatures, trying EIP-6492, EIP-1271, and ECDSA recovery in order. 

Create and pass default public clients to validation functions that may need them.

## Merge Checklist

_Choose all relevant options below by adding an `x` now or at any time before submitting for review_

- [x] PR title adheres to the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standard
- [x] PR has a [changeset](https://github.com/farcasterxyz/hub-monorepo/blob/main/CONTRIBUTING.md#35-adding-changesets)
- [x] PR has been tagged with a change label(s) (i.e. documentation, feature, bugfix, or chore)
- [x] PR includes [documentation](https://github.com/farcasterxyz/hub-monorepo/blob/main/CONTRIBUTING.md#32-writing-docs) if necessary.
- [x] All [commits have been signed](https://github.com/farcasterxyz/hub-monorepo/blob/main/CONTRIBUTING.md#22-signing-commits)

## Additional Context

Related protocol docs PR: https://github.com/farcasterxyz/protocol/pull/129

**Hubble**
- **Hub:** Create an Optimism `publicClient` at hub creation time.
- **Storage Engine:** Add `l2PublicClient`. Pass public clients through to validation functions.
- **Storage Engine Worker:** Pass serialized RPC config to worker thread, create and pass public clients to validation functions.

**Core**
- **New:** Add `eth/clients` and `eth/chains` modules, providing default mainnet/testnet public clients. 
- **Validations:** Pass provided or default public clients through validations to signature verifier. Validations that do not explicitly provide `PublicClients` will inherit the defaults.
- **Crypto:** Add smart contract signature verification using [`publicClient.verifyTypedData`](https://viem.sh/docs/actions/public/verifyTypedData.html). This function makes an RPC call to verify smart contract signatures, trying EIP-6492, EIP-1271, and ECDSA recovery in order. Note that we don't use this for EOA recovery to avoid making an unnecessary RPC call. Instead we call the "offline" [`verifyTypedData`](https://viem.sh/docs/utilities/verifyTypedData.html) util (i.e. the existing EOA signature verification behavior).
- **Protobufs:** Update `VerificationAddEthAddressBody` protobuf to add `chainId` and `verificationType`.
- **Signers:** Add optional `chainId` argument to `signVerificationEthAddressClaim`
- **Factories:** Add transient `contractSignature` parameter to verification factory.

---

## PR-Codex overview
This PR focuses on adding FIP-8 contract verifications to the Farcaster codebase. 

### Detailed summary
- Added FIP-8 contract verifications to the `eth` module
- Updated the `signVerificationEthAddressClaim` method in various signers to include the `chainId` parameter
- Added `defaultPublicClients` and `defaultL1PublicTestClient` to the `clients` module
- Updated the `VerificationAddEthAddressBodyFactory` in the `factories` module to include the `contractSignature` parameter
- Updated the `Hub` and `ValidationWorkerData` classes in the `hubble` module to include support for optimism chains
- Updated the `Message` and `MessageData` protobufs in the `schemas` module to include `verification_type` and `chain_id` fields

> The following files were skipped due to too many changes: `apps/hubble/src/storage/engine/index.test.ts`, `packages/hub-web/src/generated/message.ts`, `packages/hub-nodejs/src/generated/message.ts`, `packages/core/src/protobufs/generated/message.ts`, `packages/core/src/crypto/eip712.ts`, `packages/core/src/validations.ts`, `apps/hubble/src/storage/engine/index.ts`, `packages/core/src/validations.test.ts`, `packages/hub-web/src/generated/rpc.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: apps/hubble/src/storage/engine/index.ts [hunk 1]
```diff
@@ -1093,6 +1101,37 @@ class Engine extends TypedEmitter<EngineEvents> {
 
     return ok(undefined);
   }
+
+  private getPublicClients() {
```
Comment (adityapk00) [inline apps/hubble/src/storage/engine/index.ts:1105]: Can you add the return type to the function signature? Helps catch bugs in the future if we change what we return here.

File: apps/hubble/src/storage/engine/index.ts [hunk 2]
```diff
@@ -1093,6 +1101,37 @@ class Engine extends TypedEmitter<EngineEvents> {
...
+    if (this._publicClient?.chain) {
+      clients[this._publicClient.chain.id] = this._publicClient;
+    }
+    if (this._l2PublicClient?.chain) {
+      clients[this._l2PublicClient.chain.id] = this._l2PublicClient;
+    }
+    return Object.keys(clients).length > 0 ? clients : undefined;
+  }
+
+  private getWorkerData() {
```
Comment (adityapk00) [inline apps/hubble/src/storage/engine/index.ts:1116]: Return type (same comment as above)

File: apps/hubble/src/storage/engine/validation.worker.ts [hunk 3]
```diff
@@ -1,12 +1,39 @@
 import { validations } from "@farcaster/hub-nodejs";
 import { nativeValidationMethods } from "../../rustfunctions.js";
-import { parentPort } from "worker_threads";
+import { workerData, parentPort } from "worker_threads";
+import { http, createPublicClient, fallback } from "viem";
+import { optimism, mainnet } from "viem/chains";
+
+interface WorkerData {
```
Comment (adityapk00) [inline apps/hubble/src/storage/engine/validation.worker.ts:7]: Rename to "ValidationWorkerData" and use it as a return type above.

**Review by adityapk00 — APPROVED** [review]
I left some type-related comments, but otherwise looks good to me

File: packages/core/src/crypto/eip712.ts [hunk 4]
```diff
@@ -66,10 +77,64 @@ export const verifyVerificationEthAddressClaimSignature = async (
...
+  const valid = await ResultAsync.fromPromise(
+    client.verifyTypedData({
+      address: bytesToHex(address),
+      domain: { ...EIP_712_FARCASTER_DOMAIN, chainId },
+      types: { VerificationClaim: EIP_712_FARCASTER_VERIFICATION_CLAIM },
+      primaryType: "VerificationClaim",
+      message: claim,
+      signature,
+    }),
+    (e) => new HubError("unknown", e as Error),
```
Comment (sanjayprabhu) [inline packages/core/src/crypto/eip712.ts:106]: I assume we'll get back `false` if the message fails to validate? If so, use `unavailable.network_failure` here.

And in. https://github.com/farcasterxyz/hub-monorepo/blob/6ea579f2459df9190b98b20eb41b1bb261e88864/apps/hubble/src/storage/engine/index.ts#L397, let's make this check generic (currently only exists for username proofs) so we don't revoke valid messages if alchemy goes down temporarily.

Please add a test for this case similar to https://github.com/farcasterxyz/hub-monorepo/blob/d77970b11016226bdf1872aa6071e6de028fbd3c/apps/hubble/src/storage/engine/index.test.ts#L714

File: packages/core/src/crypto/eip712.ts [hunk 5]
```diff
@@ -66,10 +77,64 @@ export const verifyVerificationEthAddressClaimSignature = async (
...
+  signature: Uint8Array,
+  address: Uint8Array,
+  chainId: number,
+  publicClients: PublicClients = defaultPublicClients,
+): HubAsyncResult<boolean> => {
+  const client = publicClients[chainId];
+  if (!client) {
+    return ResultAsync.fromPromise(
+      Promise.reject(),
+      () => new HubError("bad_request.invalid_param", "Invalid chain ID"),
```
Comment (sanjayprabhu) [inline packages/core/src/crypto/eip712.ts:94]: I think we should have a separate check and error code for when clients are not provided vs getting an unsupported chain id. Error could be confusing for hub-nodejs consumers who don't pass in a client.

**Review by sanjayprabhu — APPROVED** [review]

**Pull request merged**